java/ql/lib/semmle/code/java/ControlFlowGraph.qll

```diff
@@ -1201,10 +1299,14 @@ private module ControlFlowGraphImpl {
       // From the entry point control is transferred first to the expression...
       n = switch and result = first(switch.getExpr())
       or
-      // ...and then to one of the cases.
-      last(switch.getExpr(), n, completion) and result = first(switch.getACase())
+      // ...and then to any case up to and including the first pattern case, if any.
+      last(switch.getExpr(), n, completion) and result = first(getAFirstSwitchCase(switch))
       or
       // Statements within a switch body execute sequentially.
+      // Note this includes non-rule case statements and the successful pattern match successor
+      // of a non-rule pattern case statement. Rule case statements do not complete normally
+      // (they always break or yield), and the case of pattern matching failure branching to the
+      // next case is specially handled in the `PatternCase` logic below.
       exists(int i |
         last(switch.getStmt(i), n, completion) and result = first(switch.getStmt(i + 1))
```

Comment:
This case should now also handle `completion = BooleanCompletion(true, _)` when `switch.getStmt(i) instanceof PatternCase`. Same goes for the identical code below for `SwitchExpr`.

Reply:
We should also add the false-edges going from `last(switch.getStmt(i), n, completion)` to the set of possible next cases when `switch.getStmt(i) instanceof PatternCase`. These are edges that go between direct children of the switch and should therefore be accounted for in this section.